QUICK-FIX Sort urls list by creation date

alieh-rymasheuski opened the pull request:
The URLs attached to an assessment are not sorted by date added.

Steps to reproduce on localhost:
- Create an assessment.
- Open its details page.
- Add an url with text 'url1' to the assessment.
- Add an url with text 'url2' to the assessment.
- With mysql client, set `creation_date`s in past for url2 and in future for url1 manually.
```sql
update documents set created_at='2000-01-01 00:00:00', updated_at='2000-01-01 00:00:00' where title='url2';
update documents set created_at='2020-01-01 00:00:00', updated_at='2020-01-01 00:00:00' where title='url1';
```
- Refresh the page.
- Add an url with text 'url3' to the assessment.

Expected result: the urls are sorted by creation dates (the order is url2, url3, url1).
Actual result: the urls are sorted by their ids or something (the order is url1, url2, url3).

This PR adds a parameter to sort `mapping-tree-view`s, so the urls list gets sorted before rendering.

plamut commented on src/ggrc/assets/javascripts/components/mapped_tree_view.js:
```diff
@@ -39,12 +40,19 @@
+        if (this.scope.attr('sortField')) {
+          mappedObjects = _.sortBy(mappedObjects, this.scope.attr('sortField'));
+        }
+        this.scope.attr('mappedObjects').replace(mappedObjects);
```
These few lines are exactly the same as in the `init` method, please consider factoring them out to a helper method.
Looks good otherwise, I like the use of `replace` here, to not lose any event bindings on the `mappedObjects` list.

alieh-rymasheuski replied:
OK, I'll try to move these lines out into a common place.

plamut commented:
Unfortunately, I was not able to reproduce the issue locally, thus I cannot verify this fix.
I did, however, observe a critical issue (script error) _"Uncaught Cannot determine the object's file name"_ (without the PR applied, so this has nothing to do with it).

alieh-rymasheuski commented:
@plamut I'll add steps to reproduce on localhost (including manual DB manipulation); the symptoms differ from those observable on the sandbox with Assessment 2442, but it was how I tested this PR.

plamut commented:
@alieh-rymasheuski That would be great, thanks!

BTW, have you observed the abovementioned script error on develop? With it, I probably won't be able to test the fix locally, and AFAIK we need to wait for the PlumJelly release branch to be merged in order to see it fixed.

hyperNURb commented on src/ggrc/assets/javascripts/components/mapped_tree_view.js:
```diff
@@ -29,7 +30,7 @@
       var binding;
 
-      _.each(['mapping', 'itemTemplate'], function (prop) {
+      _.each(['mapping', 'itemTemplate', 'sortField'], function (prop) {
```
Why do we need `sortField` here? We are passing it in as a string, right?

alieh-rymasheuski replied:
This block translates attributes in `dash-case` from the template into `camelCase` in the controller, so it translates `sort-field="something"` into `this.scope.attr('sortField', "something")`, do I get it wrong?

hyperNURb replied:
No, it actually tries to get a value as text from that field and in case it fails it tries to get it as a variable. You don't need it for your case.

plamut replied:
CanJS already does this for us. :)
If you say e.g. `<component foo-bar="baz">` and declare scope: `{... fooBar: '@'...}`, `scope.attr('fooBar')` will already have a string value of `'baz'`.

hyperNURb replied:
If you remove this line, `https://github.com/google/ggrc-core/pull/4172/files#diff-82326d5fc488307b65aadd662e1ff639R67` will work good :-)

alieh-rymasheuski commented:
@plamut I updated the steps to reproduce to what I did on localhost; I removed the link to the sandbox as the exact issue on that page on sandbox is not reproducible locally in my case.
I didn't observe such error on localhost. It probably has something to do with gdrive integration which I don't have on localhost; try testing on a newly created assessment.

hyperNURb commented on src/ggrc/assets/javascripts/components/mapped_tree_view.js:
```diff
@@ -64,9 +85,6 @@
               return mapping.destroy();
-            })
-            .then(function () {
-              return mapping.documentable.reify();
```
Why are we removing the `reify()` call? Does it have to do something with sorting?

alieh-rymasheuski replied:
`mapping.documentable` is `undefined` after `mapping.destroy()` is finished, so we get a `cannot reify an undefined` error on url deletion. This line was meant to refresh the `documentable` (the Assessment in this case) to fetch its new state from the backend, and it didn't do that because of the `undefined` error.

It has nothing to do with sorting, it is just an error I discovered while testing my new code. I can make a separate PR if you think it's a bad approach.

hyperNURb replied:
I think it's there for a reason, we can always do following:
```
if (mapping && mapping.documentable) {
     return mapping.documentable.reify();
}
return mapping;
```

or something in that manner.

Removing it doesn't fix the problem - it hides it :-)

alieh-rymasheuski replied:
I'll make a second check; it seems to me these lines never even worked because of that `undefined` error.

alieh-rymasheuski replied:
It seems the same error was already fixed in the release branch in the same way you propose: they do `if (mapping.documentable) { return mapping.documentable.reify(); }`. I'll use the same solution then.

hyperNURb commented on src/ggrc/assets/mustache/base_templates/url_list.mustache:
```diff
@@ -7,6 +7,7 @@
   parent-instance="instance"
   mapping="instance.class.info_pane_options.urls.mapping"
   item-template="instance.class.info_pane_options.urls.show_view"
+  sort-field="instance.created_at"
```
If we want to sort by `created_at` field, then we should pass in `created_at` not `instance.created_at`, since `instance.created_at` will be a date value. Or I'm missing a point here?

alieh-rymasheuski replied:
It is a date indeed. What's wrong with sorting by date?

hyperNURb replied:
```
instance.created_at = new Date('2016-7-2');
```
I think you want to sort list of instances by `created_at` field, not by date value, or I'm mistaking?

hyperNURb replied:
If it's not clear enough, in your case `sortField` will be `new Date('2016-7-2')`

alieh-rymasheuski replied:
Why do you think so? I made a breakpoint in `_sortObjects` and saw that `this.scope.attr('sortField')` was equal to `"instance.created_at"`, not a Date object. I don't even see the logic that would store a Date object into `sortField`.

hyperNURb replied:
Now it's ok :-)

hyperNURb commented on src/ggrc/assets/javascripts/components/mapped_tree_view.js:
```diff
@@ -39,14 +40,34 @@
       // We are tracking binding changes, so mapped items update accordingly
       binding.list.on('change', function () {
-        this.scope.attr('mappedObjects').replace(binding.list);
+        this.scope.attr('mappedObjects').replace(
```
If we are doing this two times already, could we abstract this to another function?

alieh-rymasheuski replied:
Does `binding.list.on('change', callback)` pass the new value of `binding.list` into the callback? If so, we can make a common callback.

hyperNURb replied:
Afaik it will return list, but please check it, so you'll have either `list` or `binding.list`

alieh-rymasheuski replied:
It passes several values, none of which is equal `binding.list`. I think refactoring these callbacks into one common callback for both events would take too much time.

hyperNURb commented:
This works as promoted so I'm merging this 👍
